go/registry/api/sanity_check.go

```diff
@@ -0,0 +1,220 @@
+		}
+		seenEntities[entity.ID] = entity
+	}
+
+	return seenEntities, nil
+}
+
+// SanityCheckRuntimes examines the runtimes table.
+func SanityCheckRuntimes(logger *logging.Logger,
+	params *ConsensusParameters,
+	runtimes []*SignedRuntime,
+	suspendedRuntimes []*SignedRuntime,
+	isGenesis bool,
+) (RuntimeLookup, error) {
+
+	seenRuntimes := []*Runtime{}
+	for _, srt := range runtimes {
+		rt, err := VerifyRegisterRuntimeArgs(params, logger, srt, isGenesis)
+		if err != nil {
+			return nil, fmt.Errorf("runtime sanity check failed: %w", err)
+		}
+		seenRuntimes = append(seenRuntimes, rt)
+	}
+
+	seenSuspendedRuntimes := []*Runtime{}
+	for _, srt := range suspendedRuntimes {
+		rt, err := VerifyRegisterRuntimeArgs(params, logger, srt, isGenesis)
+		if err != nil {
+			return nil, fmt.Errorf("runtime sanity check failed: %w", err)
+		}
+		seenSuspendedRuntimes = append(seenSuspendedRuntimes, rt)
+	}
+
+	return NewRuntimeLookup(seenRuntimes, seenSuspendedRuntimes), nil
+}
+
+// SanityCheckNodes examines the nodes table.
+// Pass lookups of entities and runtimes from SanityCheckEntities
+// and SanityCheckRuntimes for cross referencing purposes.
+func SanityCheckNodes(
+	logger *logging.Logger,
+	params *ConsensusParameters,
+	nodes []*node.SignedNode,
+	seenEntities map[signature.PublicKey]*entity.Entity,
+	runtimesLookup RuntimeLookup,
+	isGenesis bool,
+	epoch epochtime.EpochTime,
+) error { // nolint: gocyclo
+
+	nodeLookup := &sanityCheckNodeLookup{
+		nodes:           make(map[signature.PublicKey]*node.Node),
+		nodesCertHashes: make(map[hash.Hash]*node.Node),
+	}
+
+	for _, sn := range nodes {
+		// Open the node to get the referenced entity.
+		var n node.Node
+		if err := sn.Open(RegisterGenesisNodeSignatureContext, &n); err != nil {
+			return fmt.Errorf("registry: sanity check failed: unable to open signed node")
+		}
+		if !n.ID.IsValid() {
+			return fmt.Errorf("registry: sanity check failed: node ID %s is invalid", n.ID.String())
+		}
+		entity, ok := seenEntities[n.EntityID]
+		if !ok {
+			return fmt.Errorf("registry: sanity check failed node: %s references a missing entity", n.ID.String())
+		}
+
+		node, _, err := VerifyRegisterNodeArgs(params,
+			logger,
+			sn,
+			entity,
+			time.Now(),
+			isGenesis,
+			epoch,
+			runtimesLookup,
+			nodeLookup,
+		)
+		if err != nil {
+			return fmt.Errorf("registry: sanity check failed for node: %s, error: %w", n.ID.String(), err)
+		}
+
+		// Add validated node to nodeLookup.
+		nodeLookup.nodes[node.Consensus.ID] = node
+		nodeLookup.nodes[node.P2P.ID] = node
+
+		var h = hash.Hash{}
+		h.FromBytes(node.Committee.Certificate)
+		nodeLookup.nodesCertHashes[h] = node
+	}
+
+	return nil
+}
+
+// Runtimes lookup used in sanity checks.
+type sanityCheckRuntimeLookup struct {
+	runtimes          map[common.Namespace]*Runtime
+	suspendedRuntimes map[common.Namespace]*Runtime
+}
+
+func NewRuntimeLookup(runtimes []*Runtime, suspendedRuntimes []*Runtime) RuntimeLookup {
```

Comment:
Does this need to be public? Also maybe rename to something like `newSanityCheckRuntimeLookup`.